[23.0] Ensure history export contains all expected datasets

davelopez opened the pull request:
First attempt at preventing #15976



## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

davelopez committed:
Remove no longer valid TODO comment

The rocrate.zip model_store_format is covered now under the `ModelStoreFormat.is_compressed` condition since it is a zip file.

davelopez committed:
Add test case for missing dataset file

davelopez committed:
Rename add_dataset to include_dataset_file

A little bit clearer

davelopez committed:
Ensure a dataset in OK state has a valid file

when it should be included in an export process

mvdbeek commented on lib/galaxy/model/store/__init__.py:
```diff
@@ -2213,8 +2213,19 @@ def add_dataset_collection(
+    def _ensure_dataset_file_accessible(self, dataset: model.DatasetInstance) -> None:
+        is_sessionless = dataset.id is None
+        state = dataset.dataset.state
+        # Sessionless datasets are not persisted in the object store, so we can skip the check as there is no file_path associated with them.
```
That doesn't seem right as a general statement, pushing datasets to the object store is the export model store's responsibility, but that happens after calling `add_dataset`

davelopez replied:
Thank you as always Marius! That is what I needed to know.
I was quite confused as I expected it to be already persisted at this point. I will move the check to the proper place :+1:

mvdbeek commented on lib/galaxy/model/store/__init__.py:
```diff
@@ -2128,12 +2128,12 @@ def export_history(
 
             if dataset not in self.included_datasets:
-                self.add_dataset(dataset, include_files=add_dataset)
+                self.add_dataset(dataset, include_files=should_include_file)
```
I think I would verify that the dataset exists at this spot here, but not as part of `add_dataset`.

mvdbeek commented on lib/galaxy/model/store/__init__.py:
```diff
@@ -2128,12 +2128,14 @@ def export_history(
             dataset.annotation = get_item_annotation_str(sa_session, history.user, dataset)
-            add_dataset = (dataset.visible or include_hidden) and (not dataset.deleted or include_deleted)
+            should_include_file = (dataset.visible or include_hidden) and (not dataset.deleted or include_deleted)
             if dataset.id in self.collection_datasets:
```
wouldn't this include datasets that are deleted?
should this be something like
```suggestion
            if not dataset.deleted and dataset.id in self.collection_datasets:
```

davelopez replied:
Hmmm... then I guess the whole condition about collections is redundant as it will be covered by `(not dataset.deleted or include_deleted)` in the previous line.
Do you think there could be a reason to "force" include deleted dataset if they are part of a collection? If not, we can completely remove this condition.

mvdbeek replied:
`add_dataset = (dataset.visible or include_hidden)` still makes sense, you only need to include the dataset if that is true or if the dataset is part of a collection

mvdbeek replied:
> Do you think there could be a reason to "force" include deleted dataset if they are part of a collection? 

no, can't think of a reason to do that.

davelopez committed:
Shortcut condition to avoid checking the disk

If the dataset is not in OK state there is no need to check for the file_path and the cost of checking the disk.

mvdbeek commented on lib/galaxy/model/store/__init__.py:
```diff
@@ -2215,6 +2217,12 @@ def add_dataset_collection(
     def add_dataset(self, dataset: model.DatasetInstance, include_files: bool = True) -> None:
         self.included_datasets[dataset] = (dataset, include_files)
 
+    def _ensure_dataset_file_accessible(self, dataset: model.DatasetInstance) -> None:
```
```suggestion
    def _ensure_dataset_file_exists(self, dataset: model.DatasetInstance) -> None:
```

i guess accessible could be a permission level check

mvdbeek commented on lib/galaxy/model/store/__init__.py:
```diff
@@ -2128,12 +2128,14 @@ def export_history(
             if dataset not in self.included_datasets:
-                self.add_dataset(dataset, include_files=add_dataset)
+                if should_include_file:
+                    self._ensure_dataset_file_accessible(dataset)
```
```suggestion
                    self._ensure_dataset_file_exists(dataset)
```

davelopez committed:
Use better name

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

davelopez committed:
Fix condition

To avoid leaking deleted datasets that are part of a collection

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

mvdbeek commented on lib/galaxy/model/store/__init__.py:
```diff
@@ -2215,6 +2217,12 @@ def add_dataset_collection(
 
+    def _ensure_dataset_file_exists(self, dataset: model.DatasetInstance) -> None:
+        state = dataset.dataset.state
+        if state in [model.Dataset.states.OK] and not dataset.file_name:
```
It could be worth logging the object store id as well (dataset.dataset.object_store_id)

davelopez committed:
Log object store id too

mvdbeek commented on lib/galaxy/model/store/__init__.py:
```diff
@@ -2215,6 +2217,14 @@ def add_dataset_collection(
+            log.error(
+                f"Dataset [{dataset.id}] does not exists on on object store [{dataset.dataset.object_store_id or 'None'}], while trying to export."
+            )
+            raise Exception(f"Cannot export dataset: {dataset.name}")
```
Do these messages appear in the task error message ? If yes, it might be good to include the encoded dataset id in the message. If the exception message is not user visible we should include the decoded dataset id here. I think having that id will enable admins to more easily debug why they're not available.

For users it might also be a good idea to include the hid. Maybe something like

```suggestion
            raise Exception(f"Cannot export history dataset {dataset.hid}: {dataset.name} with id {self.exported_key(dataset.id)}")
```

mvdbeek commented on test/integration/test_history_import_export.py:
```diff
@@ -198,6 +199,24 @@ def test_export_history_with_discarded_dataset_to_rocrate(self):
+        assert not os.path.exists(hda["file_name"])
+
+        # Ensure the export task fails
+        with self.assertRaisesRegex(Exception, "task failed"):
```
I suppose we have no way to assert something more specific ?

davelopez replied:
I rewrote the test to use the STS instead of the file sources so we can have a bit more control of the error returned in 
7f5e97a

davelopez committed:
Include more dataset info on error

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

mvdbeek commented on lib/galaxy/model/store/__init__.py:
```diff
@@ -2215,6 +2217,14 @@ def add_dataset_collection(
+            log.error(
+                f"Dataset [{dataset.id}] does not exists on on object store [{dataset.dataset.object_store_id or 'None'}], while trying to export."
+            )
+            raise Exception(f"Cannot export history dataset {dataset.hid}: {dataset.name} with id {self.exported_key(dataset.id)}")
```
Arg, `hid` is only part of HistoryDatasetAssociation, but that should also be the only type of DatasetInstance you can have in a history ... I guess 
```suggestion
            raise Exception(f"Cannot export history dataset {getattr(dataset, 'hid', '')}: {dataset.name} with id {self.exported_key(dataset.id)}")
```
should fix the mypy error in a way that would also work for other items. Alternatively you could narrow the type with an `assert isinstance(dataset, HistoryDatasetAssociation)`

davelopez replied:
I also fixed the exported_key parameter there.

davelopez committed:
Refactor test to assert for specific error

mvdbeek reviewed (approved):
Awesome, thanks a lot!
